src/Npgsql/NpgsqlConnector.cs

```diff
@@ -254,6 +254,30 @@ internal bool PostgresCancellationPerformed
 
         internal readonly Stopwatch QueryLogStopWatch = new();
 
+        static readonly string[] CriticalFailureCodes =
+        {
+            PostgresErrorCodes.ConnectionException,
```

Comment:
Does this mean that if I kill a connection (i.e. it's holding a transaction for too long) using `pg_terminate_backend()` that it would result in the driver thinking that it needs to failover to the next server?

Reply:
Yep, it does. The problem is, postgres throws the same exception if the database is stopped.

Reply:
Or should we just consider this as something unlikely? @roji

Reply:
Hmm... I'd be conservative here - if a code *could* be generated in a situation which doesn't imply the cluster is offline, I wouldn't treat it as a critical failure code. Like @mintsoft says, there may be some legitimate uses for `pg_terminate_backend` which shouldn't cause the cluster to be treated as offline.

Note that there's a difference depending on *where* the error is caught... Any error caught when opening a connection can probably mean that the cluster is offline (though don't forget the difference between database and cluster). On the other hand, with errors caught in other situations (i.e. NpgsqlConnector.ReadMessage) we should be a bit more careful. So maybe we can treat ConnectionException as "critical failure" when it happens on opening, but not when it's in NpgsqlConnector.ReadMessage...

Thoughts?

Reply:
Whenever there is any exception while opening a connection, I do not see any reason to not treat the host as 'offline'.
Also, this is the list of errors which [pgjdbc](https://github.com/pgjdbc/pgjdbc/blob/66b1e0200c07447f8fad2e442462bb206d61be71/pgjdbc/src/main/java/org/postgresql/ds/PGPooledConnection.java#L205) treats as 'fatal'.